test role via molecule

ltalirz opened the pull request:
Add continuous integration tests via the `molecule` package that run on
Github actions.

Test log: https://github.com/ltalirz/ansible-cvmfs-client/runs/4886355529?check_suite_focus=true

ltalirz commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,65 @@
+      env:
+        MOLECULE_DISTRO: ${{ matrix.distro }}
+
+  # release:
```
This would make it straightforward to publish this role to https://galaxy.ansible.com/ (released automatically when pushing a tag).

I can remove this again if it is not desired

ltalirz commented on defaults/main.yml:
```diff
@@ -30,7 +30,7 @@ cvmfs_force_unmount: false
 # according to cvmfs_client_pv_name, cvmfs_client_vg_name, and
 # cvmfs_client_lv_name as described below. If false, you will need to
 # configure the cache storage yourself instead, and also ensure it is the required size.
-cvmfs_client_configure_storage: true
```
This is the only aspect I changed.

Reading the description of this variable gives the impression, particularly to new users who are not familiar with the CVFS, that they need to e.g. create a new file system to hand to CVFS while instead they can simply set this to `false` and the CVFS client handles everything automatically (correct?).

rptaylor replied:
That impression is generally correct; it is recommended to create a new filesystem for the cache (/var/lib/cvmfs) so that the storage for the cache is dedicated and guaranteed to be available. If it is just in / then the CVMFS client could be deprived of cache space if any other application or process on the system uses too much space.  If `true` the ansible role sets it up for you, so isn't that the easy route? It's not related to FUSE.

ltalirz replied:
If `true`, the ansible role expects a volume group to be present, i.e. it is not automatic.
If the default behavior was fully automatic this would not be an issue.

I would just point out that the [default installation instructions for the CVMFS](https://cvmfs.readthedocs.io/en/stable/cpt-quickstart.html) correspond to the option `false` here (if I understand correctly).

rptaylor replied:
No, if true it creates a PV, VG and LV for you:
https://github.com/cvmfs-contrib/ansible-cvmfs-client/blob/master/tasks/main.yml#L12
https://github.com/cvmfs-contrib/ansible-cvmfs-client/blob/master/tasks/storage.yml
You just need to tell it the physical device on which to make them, e.g. /dev/vdb, or you can use an existing VG instead as described in the documentation.
https://github.com/cvmfs-contrib/ansible-cvmfs-client/blob/master/defaults/main.yml#L35

rptaylor replied:
The quickstart guide is just that, to start quickly and get it working, suitable for a minimal basic setup , but if you're using the Ansible role you're probably installing the client on many systems or a cluster intended for production use, and on which many other applications or users could be doing things that consume disk space, in which case the cache should be on a dedicated filesystem to ensure the right amount of space is available, corresponding to the configured value of CVMFS_QUOTA_LIMIT.

ltalirz replied:
> You just need to tell it the physical device on which to make them

What if my node/VM/... in its current setup does *have* another physical device on which to make them?

In the end, all I was trying to say is that it can be helpful if the default settings of the ansible role let it run through without extra requirements, and that from the description of this variable alone it was not clear to me that this option even exists (namely by passing `false` to `cvmfs_client_configure_storage`).

I see that you have thought about the default of this value, and so I will revert this change
(it just means that users who want to give this role a "quick try" on some VM will either need to change the value of this option or mount an extra hard disk).

rptaylor replied:
> In the end, all I was trying to say is that it can be helpful if the default variables of an ansible role let it run through without extra requirements

Yes, I agree, I can see that. I thought about it more but I considered that the risk of accidentally breaking things outweighs the convenience of just working out of the box. I'm not sure if there is an easy way to have both convenience and safety, but I recently updated the var descriptions to try to explain this better.

No worries, sorry it was not clear and sorry for the long delay @ltalirz

rptaylor commented on molecule/default/verify.yml:
```diff
@@ -0,0 +1,9 @@
+  become: true
+
+  tasks:
+  - name: Test loading the Compute Canda module environment
```
This is more a test of the content delivered by CVMFS than a test of the functionality of CVMFS itself. 

The probe and chksetup are simpler and more direct, but maybe using all of them is best for testing as much as possible.

rptaylor replied:
`Compute Canda`  small typo

rptaylor commented:
@ltalirz sorry for the delay, are you still interested in this?  I'm afraid I'm not familiar with github actions or molecule but it would be useful to have working.

ltalirz commented:
Hey @rptaylor , sure if this is a welcome addition I'll have a go at getting this ready on the weekend

rptaylor commented on molecule/default/converge.yml:
```diff
@@ -0,0 +1,19 @@
+      cvmfs_client_configure_storage: false
+      cvmfs_cache_size: "1000"  # small cache size for CI
+      cvmfs_http_proxy: "DIRECT"
+      cvmfs_configuration:
```
changing cvmfs_configuration is not needed anymore and should be removed; Compute Canada repos are available by default.

rptaylor commented on molecule/default/converge.yml:
```diff
@@ -0,0 +1,19 @@
+    when: ansible_os_family == 'Debian'
+
+  roles:
+  - role: cvmfs_contrib.cvmfs_client
```
should it be `ansible-cvmfs-client` ?

ltalirz replied:
If you ever want to publish this role on ansible galaxy, you will need to switch to underscores.

Having hyphens in the repository name and underscores in the role name is common.
It is also common to include the `ansible-` prefix in the repository name and to remove it for the role name (all roles on ansible galaxy are ansible roles...).

In summary, this is the name I would suggest.

ltalirz commented:
When not explicitly providing the `CVMFS_SERVER_URL`, I now get this:

```
TASK [cvmfs_contrib.cvmfs_client : CVMFS check setup] **************************
  fatal: [instance]: FAILED! => {"changed": false, "cmd": ["/usr/bin/cvmfs_config", "chksetup"], "delta": "0:00:19.574378", "end": "2022-07-24 20:29:34.598845", "msg": "non-zero return code", "rc": 1, "start": "2022-07-24 20:29:15.024467", "stderr": "", "stderr_lines": [], "stdout": "Warning: cache limit for soft.computecanada.ca is very low (below 1GB)\nWarning: cache limit for cvmfs-config.computecanada.ca is very low (below 1GB)\nError: required parameter CVMFS_SERVER_URL undefined for cvmfs-config.computecanada.ca", "stdout_lines": ["Warning: cache limit for soft.computecanada.ca is very low (below 1GB)", "Warning: cache limit for cvmfs-config.computecanada.ca is very low (below 1GB)", "Error: required parameter CVMFS_SERVER_URL undefined for cvmfs-config.computecanada.ca"]}
```

Not sure whether this is intentional...

ltalirz commented:
After specifying `CVMFS_SERVER_URL: http://cvmfs-cache.arbutus.cloud.computecanada.ca:3128` I'm now getting

```
  TASK [cvmfs_contrib.cvmfs_client : CVMFS probe] ********************************
  fatal: [instance]: FAILED! => {"changed": false, "cmd": ["/usr/bin/cvmfs_config", "probe"], "delta": "0:00:19.474367", "end": "2022-07-24 20:50:43.493059", "msg": "non-zero return code", "rc": 1, "start": "2022-07-24 20:50:24.018692", "stderr": "", "stderr_lines": [], "stdout": "Probing /cvmfs/soft.computecanada.ca... OK\nProbing /cvmfs/cvmfs-config.computecanada.ca... Failed!", "stdout_lines": ["Probing /cvmfs/soft.computecanada.ca... OK", "Probing /cvmfs/cvmfs-config.computecanada.ca... Failed!"]}
```

Would you happen to know what might be the issue here?

rptaylor commented:
It is not normal to set CVMFS_SERVER_URL. That would cause the client to try connect to the same single server for every repository, even ones hosted by different organizations. That breaks the configuration. CVMFS_SERVER_URL has to be removed.  Besides, cvmfs-cache.arbutus.cloud.computecanada.ca is a squid proxy, not a stratum server anyway.

You also need to remove all CC config packages. Your setup is trying to open the cvmfs-config.computecanada.ca repo , but that should not be referenced anywhere.  You must still have some leftover config after removing cvmfs-config-computecanada   https://github.com/cvmfs-contrib/ansible-cvmfs-client/pull/1/commits/776798fc056768e40697856c6daaaa9fe7c82c86   ?

ltalirz commented:
Hi @rptaylor, the purpose of this pull request is to set up continuous integration test of the ansible role.
For this test it is necessary to pick a reliable server to connect to - otherwise we can't really test whether the role works (the alternative would be to spin up a separate CVMFS server inside the test environment, which would make the setup more complicated).

The choice of the server only affects test runs. When people clone the repository and _use_ the role, they will need to chose the server by themselves.
I personally don't care which server we pick - please let me know your suggestion - but we will need to pick one.

> You must still have some leftover config after removing cvmfs-config-computecanada https://github.com/cvmfs-contrib/ansible-cvmfs-client/commit/776798fc056768e40697856c6daaaa9fe7c82c86 ?

Every integration test starts completely from scratch. There is no interaction between the state of the git repository in different commits.

The outcome of the test is determined by what you see in the "files" tab, and you can see the test runs for each commit on https://github.com/ltalirz/ansible-cvmfs-client/actions
Once the PR is merged, there will be a little icon at each commit that brings you to the corresponding test run.

rptaylor commented:
Thanks for the link to the [actions](https://github.com/ltalirz/ansible-cvmfs-client/actions), that helps show what is happening.
I fixed a minor issue https://github.com/cvmfs-contrib/ansible-cvmfs-client/pull/7 that would report a spurious error from the monitoring commands, which should not use cvmfs-config.computecanada.ca anymore.

CVMFS users never need to pick a reliable stratum server to connect to. If you found that information from any documentation let me know so I can clarify it. CVMFS_SERVER_URL, and other essential information for connecting to repos, is pretty much always distributed from upstream configuration sources (via packages or config repos), regular users should not change it and don't need to worry about it. This way it just works out of the box.  (By use of a CDN, users also generally don't need to worry about selecting proxies either.)

I assure you that `CVMFS_SERVER_URL: http://cvmfs-cache.arbutus.cloud.computecanada.ca:3128` is incorrect and should be removed. That would set a proxy server to be used as if it were a stratum server, and it would attempt to find all repositories there.
You can e.g. `grep CVMFS_SERVER_URL -r /etc/cvmfs/  /cvmfs/cvmfs-config.cern.ch` to see what the correct values are, there are dozens of servers in dozens of files. 
However what is most mysterious is that this configuration is completely disregarded by the client, as shown by `cvmfs_config showconfig | grep SERVER_URL`.  It must have some built-in failsafe that ignores invalid configuration that would break all repositories, maybe related to the URL not having the right structure (e.g. `/cvmfs/@fqrn@`)

ltalirz commented:
Hi @rptaylor , thanks for the explanation!

> I fixed a minor issue #7 that would report a spurious error from the monitoring commands, which should not use cvmfs-config.computecanada.ca anymore (which would be noticed by having automated testing :) )

Ok, I've merged `master` back into the branch.
The reason I tried adding the `CVMFS_SERVER_URL` is that, [as explained above](https://github.com/cvmfs-contrib/ansible-cvmfs-client/pull/1#issuecomment-1193391498), I got an error message `"Error: required parameter CVMFS_SERVER_URL undefined for cvmfs-config.computecanada.ca"` when running the role.

I don't remember this happening before I rebased on your `master`, so perhaps it was introduced by some intermediate changes and is fixed by #7 . Let's see...

ltalirz commented:
Looks like it did :-)
https://github.com/ltalirz/ansible-cvmfs-client/actions

DrDaveD commented:
>  (By use of a CDN, users also generally don't need to worry about selecting proxies either.)

That's true for small numbers of isolated clients, but if there are a lot of clients grouped together at a site, it is still important to have a local site proxy cache.

rptaylor commented:
Yes, the original error message was better (clearer what the problem was).

Okay thanks @ltalirz !  I am only familiar with CI on gitlab; it seems more complicated on github but good to have automated testing.
